wip: adding in ambr parser for edd import

somtirtharoy opened the pull request.

wmorrell committed:
Replace tutorial link with Hector's ESE site

Each deployment of EDD has different data on the in-app tutorial pages.
Swapping out to the more updated tutorials made by Hector.

wmorrell committed:
Fix Access init when no defined replicate metadata

Still need an additional fix to ensure that the replicate system metadata
exists at startup. This will prevent the front-end from being non-responsive
because of a null-ref lookup on the replicate metadata type.

wmorrell committed:
Fix initialization of Dropzone elements

This is a strange error. Everything works fine if run locally. In the deployed
version, for some reason the TypeScript code for `new Dropzone()` throws a
TypeError of "x is not a constructor", where "x" varies depending on the
specific script loaded. Importing the dropzone module also adds Dropzone to the
window scope; so, this fix references the window scope of the class, instead of
whatever TypeScript compiles.

Note, that the content of the scripts deployed and the version running locally
are identical. It's unclear why there is any difference in behavior.

wmorrell committed:
Pin nodejs dropzone package to 5.7.2

Recent issues with initializing Dropzone elements, where browsers throw a
TypeError stating that the (minified name of) Dropzone is not a constructor,
appear to be related to several version bumps from the dropzone project. The
previous code was using 5.7.2; the current version is 5.7.6, and the author
makes note of some fairly major refactoring. Trying to revert to the previously
working version, until the issue of getting the project imported to our
environment is resolved.

wmorrell committed:
Update documentation with version migration info

Adding additional information to the documentation on migrating an EDD
deployment to the new 2.7.0 version. Expanding on the existing documentation
around using an S3 storage backend, with the specific commands to run to set up
the storage bucket(s). Adding another section on properly setting up the
configuration for imports.

wmorrell committed:
Remove call to non-existant method

The refactoring for the notification async tests, had a call erroneously added
to the connect method for import notifications. The call exists on the general
notification consumer, and is meant to populate existing notifications on
connect. The import notifications are a per-page notification, and do not
support having pre-existing notifications to display on connect.

wmorrell committed:
Show JWT login options in Swagger

wmorrell committed:
Replace old container registry with new name

We are using cr.ese.lbl.gov as a Docker image repo now. Replacing all uses of
jenkins.jbei.org:5000 with cr.ese.lbl.gov instead.

wmorrell committed:
Fix attach of click handler to permission button

This was a long-standing annoying bug on Firefox, but worked fine on Chrome and
Safari until very recently. It appears that setting the click handler on the
permission button was failing, because the function that added the handler was
wrapped in a page load event handler, which in turn was wrapped in a handler
that waited for the EDD AJAX data for the study to download. The page would
finish loading before the AJAX custom event fired, and thus the handler that
attached the click event would not run.

From a quick scan, it appears there is no longer any code in the setup for
permissions that requires waiting for the AJAX payload to finish downloading
and processing. Moving the permission setup to the top level of the script,
next to the custom event, fixes the problem. It seems that Firefox was first to
stop processing handlers for a load event if the event handler is registered
*after* the page already loaded. Chrome and Safari were continuing to
automatically trigger the handler if the page already loaded until recently.

wmorrell committed:
Switch custom profiles UserManager base class

Instead of using a plain Manager object to attach the profiles custom manager
to the User model, extend the default UserManager instead.

I noticed that some code paths would end up calling the profiles manager as if
it were a UserManager, and would throw AttributeError exceptions, because the
plain Manager is not a UserManager. Changing the base class should resolve
that problem.

wmorrell committed:
Add a timeout to Deploy Test phase

Something was causing a failure of a deploy on the websocket container in
edd-test today. The phase took over 50 minutes before I killed the pipeline.
Normal deploys take under 3 minutes.

Now there is a five-minute timeout on the Deploy Test phase when commits land
in master. If the timeout hits, an email goes to the committer so the delay can
be investigated.

wmorrell committed:
Add icon import for TinyMCE

Updated version of TinyMCE editor must include an explicit import for icons on
toolbars, otherwise buttons will appear with "not found" text instead of icons.

wmorrell committed:
Override Bootstrap 3 style for DETAILS / SUMMARY

The normalize stylesheet included with Bootstrap 3 includes adding a
`display: block` property to SUMMARY HTML elements. However, doing so removes
the triangular "twisty" icon indicating the open / closed state of the parent
DETAILS element of the SUMMARY element.

Adding a rule for any SUMMARY element that is a child of an HTML element (which
should be all of them), forcing the `display: list-item` property, and has a
higher priority than the rule set by Bootstrap. With this addition, the icons
once again show up, indicating the element can be expanded / collapsed.

wmorrell committed:
Replace hard-coded relative URL

This should be the last hard-coded HTTP URL, finishing the work on EDD-1261.
All other instances I could find have been removed during the course of
other work.

There remains a few more instances of hard-coding, related to WebSocket URLs. I
believe these can be excused for now, as Channels does not have the same
support for naming routes as Django overall does for naming paths, and the
WebSocket routes used by EDD are limited to two distinct paths, unlikely to
expand in the near-to-mid-future. For those links, `/ws/notify/` and
`/ws/load/` are de-facto names for the routes, and changing the code to remove
them does little to improve the code.

wmorrell committed:
Refactor study links rendering in admin

While searching for hard-coded URLs remaining in EDD, found some code in the
admin that was still using %-formatted strings for displaying links to studies
that reference a particular object. Refactored the code into a function, and
use string.format() instead of %-format.

wmorrell committed:
Changing references to default branch name

The default branch is renamed to `trunk` from `master`. Replacing uses of the
old default name with the updated name.

wmorrell committed:
Remove doc ref to non-existant script

The bulk line creation script no longer exists in the repo. Removing the
paragraph in documentation referencing it.

wmorrell committed:
Remove rest-client Dockerfile

The code referenced in the Dockerfile definition was removed some time ago, and
the Dockerfile itself is not maintained. Removing it from the repo.

dependabot[bot] committed:
Bump django from 2.2.16 to 2.2.18 in /docker/edd/core

Bumps [django](https://github.com/django/django) from 2.2.16 to 2.2.18.
- [Release notes](https://github.com/django/django/releases)
- [Commits](https://github.com/django/django/compare/2.2.16...2.2.18)

Signed-off-by: dependabot[bot] <support@github.com>

wmorrell committed:
Re-lock all dependencies

Dependabot is doing these one at a time. I do not want to wait on a piecewise
approach, so just updating everything at once. Can roll back and investigate
further if anything breaks.

wmorrell committed:
Fix blank modal in empty study

Fix for EDD-1340. When navigating to the Lines view of a Study, the function to
set up the page modal dialogs was tied to the loading of the Handsontable
widget. Since the table never loads when there is no line data, the modals are
not properly set up in an empty Study. Moving the modal setup to the callback
for page loading allows for all the correct setup on the modal, regardless of
whether the table loads.

chimaerase committed:
Fixed import documentation

EDD-1339: Drill-downs for available protocols & measurement units were broken.

wmorrell committed:
Add a limit to values displayed in graphs

On some studies with large amounts of data, the performance of showing graphs
can degrade significantly. Limiting the number of values displayed at a time
mitigates the problem. Fixes EDD-1352.

wmorrell committed:
Improve build process for edd-core Dockerfile

A handful of small tweaks to how the Dockerfile for the edd-core image is
building. The result of these tweaks are a final image that's about 40MB
smaller, and builds on my laptop around 25-30 seconds faster.

Adding an additional build argument, `REPO_SERVER`, for specifying which
container registry to use for base images. The default is set to "docker.io",
which is the same as just leaving it blank. This will let our automated builds
specify our local registry instead during builds, to avoid needing to pull or
push from the public "docker.io" container registry.

The build phases are now more strictly separating the TypeScript builds from
the Python code. There should no longer be a need to wait for TypeScript to
rebuild if only Python code is changing, which will save significant time;
TypeScript builds typically take over a minute. An additional build phase is
added to load in the current git commit hash, which was previously being done
along with the TypeScript build. The Python code is now loaded in directly,
instead of being passed along from the TypeScript build phase. This means that
Python-only changes can now be rebuilt in a matter of seconds.

wmorrell committed:
Rename Category to MeasurementClass

The name "Category" is too vague, and clashes with the name Category as used in
imports for the edd.load package. The interface is meant to combine the values
of a measurement type with its compartment for metabolites, so renaming to
better reflect that.

wmorrell committed:
Fix display of non-metabolites with compartment

The filter code was making assumptions that all non-metabolite measurements
would always have a "0" or "N/A" value set on compartments. This turns out to
not be the case. We still only distinguish between compartments for
metabolites, but now a measurement with a non-N/A compartment will properly
show up in graphs.

wmorrell committed:
Fix local account to LDAP conversion

Previous code was attempting to overwrite existing local password with an
invalid hash, but the save() call was passing only the string "password" to
update_fields kwarg, when it should pass a sequence type of strings containing
"password". The string itself gets interpreted as several single-character
fields as shown below:

    main.admin   ERROR    User migration to LDAP account failed The following
    fields do not exist in this model or are m2m fields: o, a, r, d, s, w, p

wmorrell committed:
Fix spurious pending migration in profile app

Django detects a migration in profile, but the migration it runs generates a
SQL no-op. Updating the existing migration for the field to have all the values
Django expects, to clear the issue.

wmorrell committed:
Add ability to edit type_group in Category admin

Fixes EDD-1355, EDD-1356. Allows admin site edits of the type_group field of
Category objects in the load app. This is needed because the value of the
type_group field of the selected Category in an import can affect the behavior
of type lookups.

wmorrell committed:
Upgrading to Django

Moving code to use Django 3.2.x and Channels 3.x. These dependencies were the
only ones pinned to older versions, unpinning Channels and repinning Django to
be a version below Django 4.0 brings the packages up to latest releases.

wmorrell committed:
Fix test expected output

Previous versions of Django would render a single-quote `'` as `&#39;` with
entity encoding. Now newer Django versions will render with the hex code,
`&#x27;` instead.

wmorrell committed:
Replace deprecated function alias

The `force_text()` function is an alias for `force_str()`. The alias is
deprecated and slated for removal in Django 4.0. See:

https://docs.djangoproject.com/en/3.1/releases/3.0/#django-utils-encoding-force-text-and-smart-text

wmorrell committed:
Adding simplified Strain Registry API

Including an updated and simplified API to interact with ICE, as part of the
edd.search app. Replacing integration tests using the old API with updated
tests using the new API; this is cutting down significantly on the number of
warnings emitted for hanging sockets after running the test suites.

wmorrell committed:
Replace patched cleanse_setting with new API

Django prior to 3.2 used a function `django.views.debug.cleanse_setting()` to
replace any potentially sensitive parameters in settings with an obfuscated
value (a string of asterisks). In Django 3.2+, this function is replaced by a
SafeExceptionReporterFilter class, that handles cleansing settings, as well as
sensitive parameters in request META headers and post parameters, when
reporting values in errors.

EDD was using a monkey-patched version of `cleanse_setting`, that checked for
keys likely to contain URLs, then parse those URLs and replace any found
passwords with asterisks. Now that the original function is gone and refactored
into a class swappable in settings, moved the patched code into a sub-class of
the new class, and changed the base EDD settings to load our version of
the class.

wmorrell committed:
Fix error removing unset permission from campaign

It's unclear why this error was not happening before, it could be an issue with
ordering of tests. When removing a permission from a campaign, if the
permission did not exist before, it should be a no-op.

wmorrell committed:
Fix deleting of lines

Django as of 3.2 dis-allows doing QuerySet .delete() calls after calls to
.distinct(), because in some cases there are reports of .delete() after
.distinct() generating SQL like `DELETE FROM table;`. Deletion of lines is
making use of the ExportSelection class, that can take identifiers for studies,
lines, assays, and measurements, and create a selection that includes all
parent objects for the given identifiers, and all child objects. This ends up
using a call to .distinct() for each level. Getting around the limitation by
selecting the id values for lines, then doing a delete on lines having id in
that set.

wmorrell committed:
Fix test using abstract ORM class

Django as of 3.2 dis-allows instantiating an ORM class that has a Meta class
defining abstract to True. Instead of using the abstract EDDMetadata class for
testing, instantiate the concrete EDDObject instead.

wmorrell committed:
Fix styling clash with Django admin in dark mode

The new Django versions include dark mode support in the admin. Our custom
scripts and styling clash with the dark mode styling. Re-arranging the scripts
to only appear on our handful of custom pages restricts the display weirdness
to just those few pages, instead of the entire admin site.

wmorrell committed:
Replace contrib JSONField with Django JSONField

Django now has bundled support for a JSONField on models, and the same field in
the django.contrib.postgres package is now deprecated. The contrib version will
be removed in Django 4.0. See:

https://docs.djangoproject.com/en/3.2/releases/3.1/#deprecated-jsonfield

The migration generated from the implementation switch is a SQL no-op, so
modifying the existing migrations instead of adding further no-op files to the
migration history.

wmorrell committed:
Improve message for limiting values in plots

The message displayed when a plot has stopped adding additional values to
display on the Study Data view has some updated wording to make it clearer what
is happening. It's also made more prominent by using the warning alert class,
so it shows as a larger yellow warning banner instead of just a line of text.

wmorrell committed:
Add spacing for export button in Lines view

On production deployments, the "Report a bug" issue collector from Jira was
getting rendered near the spot where the Export button is placed, partially
obscuring the button. Adding some additional spacing to account for the 24
pixel height of the issue collector button, and clarified the components of the
fudge factor used to calculate the sizing of the table and positioning of the
controls on the bottom of the table.

wmorrell committed:
Change y-axis extent calculation

The previous calculation made assumptions about wanting a zero-bound y-axis,
which may not hold. It also forced the difference between max and min to be at
least one.

The new calculation does not force the y-axis to contain zero, and only forces
the difference between max and min to be one when the actual difference
is zero.

To demonstrate:

Range: 6.85 to 7.12
Old axis: 0 to 7.12
New axis: 6.85 to 7.12

Range: 0.021 to 0.054
Old axis: 0 to 1
New axis: 0.021 to 0.054

Range: 0 to 0
Old axis: 0 to 1
New axis: 0 to 1

Range: 42 to 42
Old axis: 0 to 42
New axis: 42 to 43

wmorrell committed:
Change bar graph width to use window width

The limits to number of bars shown in a bar plot was using the coordinate space
width of the overall plot, so it was always 250. Make it instead go off the
pixel width of the window, so resizing will add or remove bars as the window
size changes.

somtirtharoy committed:
wip: adding in ambr parser for edd import

wmorrell committed:
Normalize export button between views

The "Export Data" button available on the Description (Lines) view, and the
Data view, was inconsistent. Only the Description view had options on the
button, while the Data view only supported the CSV export. Adding in the menu
options available from the Description view version of the button to the Data
view version.

chimaerase committed:
Fix styling in load success email

chimaerase committed:
Add documentation for ICE folder search performance configuration

somtirtharoy committed:
{wip} Fixed pre-commit errors

chimaerase committed:
Split edd.describe tests into separate files

Breaking up existing tests into separate files according to what's being tested.  Preparing for addition of many more new tests.

chimaerase committed:
Replace core edd.describe exceptions

Preparing for use of new error reporting code, and accompanying addition of new exceptions that cover all of the current detected error conditions.

1. Add new core exception classes for use in refactored edd.describe code
   Copied MessagingMixin and related tests in test_reporting.py from edd.load.exceptions.core.
2. Move and rename the single existing exception DescriptionError to edd.describe.exceptions.core.py::DescribeAbortError to better capture its purpose, and to differentiate from more generic DescribeError parent class added here

chimaerase committed:
Add core error reporting code to edd.describe

Add more mature reporting code and tests copied from edd.load for use in simplifying edd.describe's error reporting. This code may eventually be centralized if it penetrates more deeply into the EDD code and perhaps evolves a bit more first. Copying due to wmorrell's feedback

chimaerase committed:
Add Redis-based error tracking for edd.describe

The edd.describe.broker.DescribeErrorReport class added here (lifted from edd.load) supports the immediate need to refactor email-based reporting of errors and warnings, and will also likely be useful as a basis for later, wider use of Redis similar to edd.load’s LoadRequest.  Also lifted the subset of relevant tests from edd.load.  Likely in the future that existing CombinatorialDescriptionInput, ExperimentDescriptionOptions, and this new DescribeErrorReport may be combined to further simplify code (with much wider scope of changes). Purposefully deferring that change until a later round of refactoring to limit the scope of code changes in this round.

chimaerase committed:
Add new email reporting code for edd.describe

Replaces string-based email construction with use of Django templates, which enable translation/localization, and also separates most of email construction code from the main codebase.  Purposefully preferring the “add lines” language, e.g. “adding lines to [study] failed” vs something more generic like “describe experiment failed”, since adding lines is the only functionality supported ATM, and also matches the “add line combo’s” GUI’s labeling and final “add lines” button.   We’ll need to update this, e.g. when that GUI gets revisited, or when the ability to edit existing lines is added.  This language may be a little confusing for API users of /rest/study/{id}/describe/, but probably not too much, and makes sense to prioritize UI users at this point.

somtirtharoy committed:
{wip} removed redundant code and added commants for better readability

wmorrell committed:
Update prettier formatter to 2.3.0

The build was already grabbing the latest prettier 2.3.0 during the build of
the edd-core image. There are some changes to how prettier will wrap long
lines, which are causing a few places in the code to re-wrap.

See: https://prettier.io/blog/2021/05/09/2.3.0.html

wmorrell committed:
Lock prettier version to 2.3.0

To prevent any further issues with builds breaking because our formatter
updates -- see previous commit -- lock the version of prettier in the node
image to 2.3.0 to match the version in pre-commit.

wmorrell committed:
Re-lock Pipfile to fix test launch

A regression in Python 3.8.10 is causing the loading of the faker module to
fail. The project has released a fix with faker version 8.1.3 that should not
trigger the upstream regression.

This is also updating several other dependencies, as I ran `pipenv lock` rather
than updating just the individual testing dependency. There are a few pending
updates to Django, Pillow, etc anyway, so this commit will resolve
those updates.

wmorrell committed:
Remove deprecated/unused staticfiles template lib

The 500.html error page template is loading "staticfiles". Nothing in the page
is using the template tags provided by that module, and that name for it is
deprecated in favor of "static" instead.

wmorrell committed:
Use container static files as default fallback

Building the EDD image will bundle in static assets with the Docker image. When
running in development with volume-mounted code, there will often not be any
static assets saved in the working development directory. This is causing a
fresh clone and startup to fail, as EDD will try to read static assets from the
volume-mounted directory, and nothing will be there in a fresh clone. Setting
the bundled version as an additional location with STATICFILES_DIRS will allow
those files to be found, and for startup to complete.

wmorrell committed:
Include localhost in default ALLOWED_HOSTS setting

The previous default of an empty list for ALLOWED_HOSTS was causing health
checks on the http and websocket containers to fail, because Django returns a
500 response for a request not on an allowed host. Adding this change allows
EDD to work correctly from a fresh clone as described in the documentation.
Real deployments should still initialize code in the local settings module and
assign real host strings to the ALLOWED_HOSTS setting.

wmorrell committed:
Remove references to upload size limits

EDD no longer is storing files on local disks, and is instead using S3 buckets.
This, combined with using Traefik for a proxy, eliminates the technical
requirements to have upload file size limits and defaults. A few places in the
code were still referencing these limits, and are now removed.

somtirtharoy committed:
Add parser for parsing Ambr250 data for import into EDD

{wip} put ambr parser in a separate file, created new mixin for multi sheet ecel parsing, setup ambr parser testing code

{wip} added pandas in Pipfile and Pipfile.lock, put Ambrparser in its own file

Added mapper and codeto manipulate the ambr data using the GenericImportParser, wrote a simple test to check if parsed result is valid, addded Pipfile to root directory dso that tests can be run locally using IntelliJ, created a demo.py for testing

{wip} modified multisheetexcelparser to combine all imports as a single import call to the GenericImport parser parse_rows, modify the Measurement Mapper to have a map_data method that does alll the work, add drop nan values form the data

Made changes according to feedback from pull request

Modified test for ambr parser, removed large file from data folder

Added condition for renaming measurement types for fidelity with EDD

Add Layout for Ambr parser, modify handling of NaN values in the parser, make necesssary changes in testing the parser, rename sheet names in the Ambr test file

{wip} Add DefaultUnit class to make add mapping between measurement types and measurement units

wmorrell committed:
Fix loading of ASGI application

The upgraded Channels library makes some changes to how ASGI applications are
loaded. Updating our code to more closely match what the 3.x documentation
shows. Wrapping the imports of EDD custom code and AuthMiddlewareStack in a
function, as the default example shown in the documentation will have imports
re-organized in a way that will break the application.

wmorrell committed:
Add Django DEFAULT_AUTO_FIELD setting

With Django 3.2, the auto-generated primary key field on models is
configurable. The previous default would use an IntegerField (32-bit) under the
hood, while the new default uses a BigIntegerField (64-bit). With this switch,
there are now migration warnings. Explicitly adding a value of AutoField uses
the old behavior, and does not result in a whole-database migration.

See: https://docs.djangoproject.com/en/3.2/releases/3.2/#customizing-type-of-auto-created-primary-keys

wmorrell committed:
Fix parsing of files with BOM

Excel will sometimes save CSV files with a Unicode Byte Order Marker, using the
zero-width non-breaking space character, which shows up in Python as "\ufeff".
When EDD gets a file like this, the BOM is not stripped from the string,
causing a failure to match to expected column headers. This adds a test for the
situation where there's a BOM in a file, and changes the code to pass the test.

wmorrell committed:
Install pygments in doc builder

With the pygments package installed, the generated documentation can have
better syntax highlighting of code blocks.

wmorrell committed:
Add documentation content from ESE Google Site

Folding in a big chunk of documentation content that currently exists in the
Google Site for ESE Data & Automation. This also includes a bit of
restructuring in the MkDocs table of contents configuration.

wmorrell committed:
Fix collectstatic script in development

After the streamlined build changes in
66dabaa78eac1ddb066672674b15a3b7cbde63ef, where there's a cleaner separation of
the TypeScript build from the Python code, and the update in
171869d1d41a1c412c3e485f8442d21b4ada9fdd to use the container copy of static
files as a fallback, using the refresh_static.sh script in development would
not pick up local changes in Docker volume mounted code. By tweaking the config
used to do the static files collection, local development functions smoothly
again, keeping separate configuration from normal production, startup,
and development.

wmorrell committed:
Fix AutocompleteRenderer for Assay experimenter

The form field renderer for an assay experimenter was using an outdated
refactoring, where it was expecting to get an EDDContact in the render
callback. The final refactoring actually sends an argument of the AssayRecord.
The code is now properly dereferencing the callback parameter and providing the
correct information for rendering form elements.

wmorrell committed:
Add edit_uri key to mkdocs.yml

The default MkDocs implementation assumes that all GitHub repos use "master" as
the default branch. We use "trunk" as the default branch, and the "master"
branch no longer exists. This assumption causes the link for "Edit in GitHub"
to return a 404 page, as the link goes to a branch that does not exist.

Changing the edit_uri key to explicitly use a path to edit in "trunk" fixes the
problem. See [the edit_uri docs][1], specifically the note section that says:

    For example, for a GitHub- or GitLab-hosted repository, the `edit_uri`
    would be automatically set as `edit/master/docs/` (Note the `edit` path
    and `master` branch).

    …

    To use a different URI than the default (for example a different branch),
    simply set the `edit_uri` to your desired string.

[1]: https://www.mkdocs.org/user-guide/configuration/#edit_uri

wmorrell committed:
Change line "Disable" to "Delete"

Following recommendation from Truss; the mixing of "Disable" and "Delete"
language, and using iconography tied to delete (trash can icon), is confusing.
Simplifying the presentation to always use "Delete" instead of "Disable". The
backend functionality remains the same, if a Line has no Assays or Measurements
linked, it gets truly deleted; otherwise, the "active" flag is set to False to
disable the line and hide it from the normal interface.

dependabot[bot] committed:
Bump django from 2.2.22 to 2.2.24 in /docker/edd/core

Bumps [django](https://github.com/django/django) from 2.2.22 to 2.2.24.
- [Release notes](https://github.com/django/django/releases)
- [Commits](https://github.com/django/django/compare/2.2.22...2.2.24)

---
updated-dependencies:
- dependency-name: django
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

wmorrell commented on server/edd/load/parsers/core.py:
```diff
@@ -641,6 +643,104 @@ def _raw_cell_value(self, cell):
         return val
 
 
+class MultiSheetExcelParserMixin:
```
This is a bad name, given what the class actually does. If you're going to make a mixin class for handling multiple sheets, the class should *only* concern itself with handling multiple sheets. It should delegate responsibility for interpreting the *contents* of those sheets to a different class.

somtirtharoy replied:
Okay I will think about how to reorder this code so this makes sense or else I will just rename the class.

wmorrell commented on Pipfile:
```diff
@@ -0,0 +1,52 @@
+[[source]]
```
Delete this file. The `Pipfile` used to install packages lives next to the `Dockerfile` for the core image, under `docker/edd/core/`.

somtirtharoy replied:
I did this to run my tests locally. I used IntelliJ which was looking for this file in the root directory. It enabled it to create a virtual environment where I could run the tests locally from IntelliJ without having to exec in to a container. I felt that it would be easier to have this in the root directory, but I can remove that.

wmorrell commented on Pipfile.lock:
```diff
@@ -0,0 +1,1553 @@
+{
```
Delete this file. The `Pipfile.lock` used to lock packages lives next to the `Dockerfile` for the core image, under `docker/edd/core/`.

somtirtharoy replied:
Same as above. Removing it.

wmorrell commented on server/edd/load/parsers/ambr.py:
```diff
@@ -0,0 +1,6 @@
+from .generic import GenericImportParser
+
+
+class AmbrExcelParser(MultiSheetExcelParserMixin, GenericImportParser):
```
Ambr does not use the Generic spreadsheet layout. Mixing up unlike things this way will only cause bugs and headaches down the line.

somtirtharoy replied:
I am using _parse_rows, _parse_row and subsequent validation functions from the GenericParser hence the inheritance. I would like to discuss this with you further. I am converting the multi sheet format to a format that can leverage the GenericImportParser framework without having to reinvent everything.

wmorrell commented on server/edd/load/parsers/core.py:
```diff
@@ -641,6 +643,104 @@ def _raw_cell_value(self, cell):
+    def __init__(self, sheet_name, df):
+        self.loa_name = sheet_name
+        self.df = df
+        self.units = {
```
Do not hard-code these. All of this must be configuration in the database. The example file you've been using is *not* complete, there are hundreds of possible columns that could be included in any given import. When you hard-code it instead of looking up values from database configuration, you're forcing a code change whenever someone decides to check a slightly different column when exporting from the instrument.

somtirtharoy replied:
Yes I will make the changes.

somtirtharoy replied:
But I don't still understand how to configure the database and would like some pointers in that direction.

wmorrell commented on server/edd/load/parsers/core.py:
```diff
@@ -641,6 +643,104 @@ def _raw_cell_value(self, cell):
+        self.df["Units"] = self.units[mtype_name]
+        # dropping records with NaN values
+        self.df = self.df[self.df["Value"].notna()]
+        self.parsed_df = self.df
```
It makes no sense to have two attributes with the same value. You want a boolean flag for whether the dataframe is "parsed" or not, instead of relying on another attribute having a dataframe or `None`.

somtirtharoy replied:
Will make the change

wmorrell replied:
This is still here. Just use `df`, `parsed_df` does not need to exist.

wmorrell commented on server/edd/load/parsers/core.py:
```diff
@@ -641,6 +643,104 @@ def _raw_cell_value(self, cell):
+        for name, sheet in wb.items():
+            # for every two columns in the worksheet
+            # corresponding to each measurement type in the sheet
+            for i in range(0, int(sheet.shape[1]), 2):
```
This loop should be the core of an Ambr parser class. First column sample name, mapped to either an existing assay or a created assay linked to a line. Then pairs of columns for values of a specific measurement and the time of that measurement.

somtirtharoy replied:
Yes I agree.

wmorrell commented on server/edd/load/parsers/demo.py:
```diff
@@ -0,0 +1,165 @@
+import numbers
```
What is the purpose of this file being stand-alone? Just turn it into normal test code and run it with the normal test facilities. Look up the `-k` flag of `pytest`.

somtirtharoy replied:
This file was my prototype file, removing this.

wmorrell commented on server/edd/load/parsers/generic.py:
```diff
@@ -95,3 +95,5 @@ class GenericExcelParser(ExcelParserMixin, GenericImportParser):
 
 class GenericCsvParser(CsvParserMixin, GenericImportParser):
     pass
+
```
You have not installed the `pre-commit` hooks. 

1. Run `pre-commit install` from the root of the repo;
2. Then, run `git fetch origin trunk:trunk`;
3. Rebase to `trunk` with `git rebase -i trunk`, and change all the beginning-of-line `pick`s to `edit`;
4. Step through the rebase process following the command line instructions, re-committing each commit with `pre-commit` enabled this time. You'll likely have to fix a few things that `pre-commit` doesn't fix for you.

somtirtharoy replied:
Okay I will go through this.

wmorrell commented on server/edd/load/tests/test_ambr_parser.py:
```diff
@@ -0,0 +1,65 @@
+    path = ("ambr_export_test_data.xlsx",)
+    parser = parsers.AmbrExcelParser(uuid4())
+
+    with factory.load_test_file(*path) as file:
```
Why are you doing argument unpacking here? Just use the path variable, not in a tuple, directly.

somtirtharoy replied:
Okay.

wmorrell commented on server/edd/load/tests/test_ambr_parser.py:
```diff
@@ -0,0 +1,65 @@
+
+    with factory.load_test_file(*path) as file:
+        parsed = parser.parse(file)
+    print(parsed)
```
Don't commit tests that use `print`. It's fine to fall back to printing things when trying to debug, but remove it before you commit.

somtirtharoy replied:
Sure, will do that.

wmorrell commented on server/edd/settings/base.py:
```diff
@@ -273,8 +273,8 @@ def __contains__(self, address):
-with open("/edd.hash") as f:
-    EDD_VERSION_HASH = f.read().strip()
+# with open("/edd.hash") as f:
+EDD_VERSION_HASH = 1287461
```
Revert this change. You're hard-coding a specific commit hash used to version the running code. With this change there's no way to easily tell which version of the code you're running anymore.

somtirtharoy replied:
Yes this was to make the tests work as it was unable to find the edd.hash file locally. I will undo this. Thanks

wmorrell reviewed (changes requested):
This needs a lot of work still.

chimaerase committed:
Remove duplicate "add line combos" option for replicates

Remove the duplicate "Replicates" option that appears in the "add line combos" GUI after recent improvements to EDD's replicate tracking / supporting addition of a new "Replicates" Line MetadataType.

1.  Updated the UI to omit the new MetadataType in favor of existing support for tracking the number of replicates to create for each line.
2.  Updated "Replicates" input label to "# Replicates" to match user feedback from this week, as well as resulting planned changes to the analogous Experiment Description file column header

wmorrell commented on server/edd/load/tests/test_ambr_parser.py:
```diff
@@ -1,21 +1,18 @@
     parser = parsers.AmbrExcelParser(uuid4())
 
-    with factory.load_test_file(*path) as file:
+    with factory.load_test_file(path) as file:
```
You took off the splat operator, but left `path` at a 1-tuple. Line 9 should be `path = "ambr_export_test_data.xlsx"` (note no parenthesis and comma, so making a string literal and not a 1-tuple).

wmorrell reviewed (changes requested):
You've added an excel hidden file `~$ambr_big_data.xlsx`. That should get removed.

wmorrell committed:
Add explicit dependency on chardet

The `chardet` dependency was previously implicit, but it is getting used in the
old `hplc.py` module. Adding the explicit dependency until that module can be
re-evaluated. Presumably, the dependency that was pulling in `chardet` as a
transative dependency moved away from it for a reason.

wmorrell committed:
Fix resolving of ICE proteins already seen by EDD

The process of resolving an ICE part ID to a protein record in EDD was always
trying to create a new protein record every time. Now there's an additional
check to see if EDD has referenced that ICE part before, and uses the
already-existing protein record from that previous reference.

wmorrell committed:
Squash profile app migrations

Doing this will eliminate use of the old HSTORE method of storing a dictionary
type in a column. Use of HSTORE was replaced with a JSONB column, but the
migrations still reference HSTORE, and require any new database deployments to
have the HSTORE module enabled.

wmorrell committed:
Fix links to UniProt

We were using an ebi10.uniprot.org subdomain for rendering links to UniProt.
The configuration of UniProt servers has changed, such that the subdomain is
getting redirected to www.uniprot.org instead. In doing this, the server is
misconfigured to omit the initial path separator between the domain and the
path. Switching to using www links should avoid this URL rewrite.

wmorrell committed:
Update setup to use Postgres 11

We're using Postgres 11 now in production, so it only makes sense to update
development environments to use it as well. Eliminating the custom container,
and instead using the official Postgres container. Any existing development
deployments can continue using a compose file with Postgres 9.6; or, update to
use Postgres 11 on a new volume.

Changes to make in compose override file:
-   Switch `postgres` and `ice_db` keys for `image` to `library/postgres:11`
-   Add environment for `POSTGRES_DB=edd` and `POSTGRES_USER=edduser`
-   Change `postgres.env` to use value of `EDD_PGPASS` for `POSTGRES_PASSWORD`

wmorrell committed:
Rebased master onto import-ambr-data branch

The provided example settings file contained a commented-out value referring to
an outdated location for the LocalTestBackend authentication backend class.
Also missing, an example of using the console backend for emails.

wmorrell committed:
Refactor nested router in REST API

The nested routers are printing warnings like this in the logs:

WARNING  /usr/local/lib/python3.9/site-packages/django_filters/rest_framework/backends.py:127: UserWarning: <class 'edd.rest.views.StudyLinesView'> is not compatible with schema generation

So, removing these from the code, rather than spending time tracking down the
issue. There is still one Viewset using a nested router, in the import code;
however, this is not reporting the above warning.

Also removing some outdated comments, and unused fields (metabolic map, carbon
source). Adding in a field for the study ID in the core model serializers.

wmorrell committed:
Add custom content negotiation to exports

Trying to run an export from the Swagger UI at /rest/docs/ would always fail,
because the Swagger UI is hard-coding an Accept HTTP header with
application/json, and there isn't a clear way to configure it to do otherwise.
The custom content negotiation will forcefully add text/csv to the Accept
header before it reaches the export CSV renderers, allowing the UI to run
exports without error.

wmorrell committed:
Add in_study parameter to REST filters

Using in_study as a parameter in REST calls will allow for using a study ID,
UUID, or slug value, when querying items. This should be a nice short-cut for
clients that are currently querying the study endpoint with the slug, then
parsing out the ID, and in turn passing that along to the API endpoint the
client actually wants.

wmorrell committed:
Add help texts to REST API fields

wmorrell committed:
Add ice database options to config generator

Replacing the hard-coded static database password for development ICE
deployments, with a generated password. The alias for the database container
for the ICE database stopped working, so adding a more dynamic configuration
became necessary. With a dynamic config added, there is no more justification
for having a hard-coded password.

wmorrell committed:
Add staticfiles collection build phase

Splitting the collectstatic command into its own phase allows for omitting the
Django app static directories in the final build, making the image slightly
smaller, and eliminating the need to do a double scan for static assets during
application launch. All the static files are already collected in the
/usr/local/edd-static folder during the Docker image build, and only that
location needs to run collectstatic at launch, to detect any updated static
assets in the image.

wmorrell committed:
Fix CI config of ICE blast install

The new version of ICE does not set a default configuration in the database for
BLAST_INSTALL_DIR, so the previous initialization script that tries to UPDATE
that in the database will not stick. Switching the script to instead INSERT the
config key.

wmorrell committed:
Fix sporadic failures from list ordering mismatch

Some parts of the edd.load module were using lists as parameters, derived from
dictionaries or querysets, without explicit ordering. When testing on these,
sometimes the ordering of the lists in the tests would differ from the ordering
returned by the code. Since these items are semantically sets, and not lists,
switching to use sets instead. Other parts of the code using the same parameter
are already converting to set, so this is making the code overall more
consistent. The test failure looked like the below:

E       AssertionError: assert {'conflicted_...s': True, ...} == {'conflicted_...s': True, ...}
E         Omitting 9 identical items, use -vv to show
E         Differing items:
E         {'loa_pks': [144, 143]} != {'loa_pks': [143, 144]}
E         Use -v to get the full diff

Note: at some point, should refactor this part of the code to use an actual
defined data structure, instead of relying on arbitrary dict objects with
specific keys. This is a relic of the previous, pre-refactor version in
edd_file_importer, that I just didn't get to properly rewriting.

wmorrell committed:
Add more test coverage to REST API

Module is now covered almost completely. The core endpoints for querying
studies are covered, with some missing tests still for assays, measurements,
measurement types, and users.

wmorrell committed:
Fix references to REST API in TypeScript

The endpoints for `measurement_types` and `measurement_units` were renamed to
`types` and `units`, respectively. Updating the TypeScript code to use the
correct URLs to access this data.

wmorrell committed:
Remove metadata_group REST endpoint

Having a distinct endpoint for the groups served no purpose. The group names
were not tied to the IDs that were displayed in metadata_types, making both
endpoints less useful. Instead, metadata_types now directly includes the group
name, if there is one.

wmorrell committed:
Add Django Signals for efficiency metric milestone

The milestone for EDD specifically states:

«DBTL cycle (wall time; capacity; unit operations) efficiency metrics data
gathering sourced from … EDD»

Adding in definitions for signals that EDD can send to begin collecting metrics
information for use of EDD features. Including additional documentation on the
already existing signals, now that the signals are serving multiple purposes.

wmorrell committed:
Change to custom user model

Replacing the monkey-patched functions on the User object with a real custom
User class that extends the default Django AbstractUser class. This makes the
application setup slightly simpler, because we don't need to hook in to add the
monkey-patched functions at the right time anymore; and Django will no longer
detect pending migrations for the django.contrib.auth app anymore because of
our custom Manager object that fetches the Profile associated with a user.

Changes here are roughly following the outline given in:
https://www.caktusgroup.com/blog/2019/04/26/how-switch-custom-django-user-model-mid-project/
and
https://code.djangoproject.com/ticket/25313

Since we don't care much about the table name in the database, using the same
auto-generated name in the standard Django model is fine. Since we also have a
freshly-squashed migration in the edd.profile app, and that app also deals with
User-related things, adding in the definition for the new User model there
allows for everything to continue functioning exactly as before for existing
installations, and still work for brand-new installations.

wmorrell committed:
Fix login failure when Solr is unresponsive

Adding a try-catch and logging a warning whenever the code updating the Solr
indices fails.

Whenever a user logs in, the database record for that user gets updated with
the last login time set to the current time. When the database record updates,
that triggers an update to Solr. But, if Solr is unresponsive or down, the
index update will throw an exception. If that exception is not caught, it
bubbles up until it ultimately fails the login request.

Fixes EDD-1362.

wmorrell committed:
Refactor permission change signal receiver

Using an if-condition-return control flow is silly when it's simpler to use an
if-not-condition-act instead.

wmorrell committed:
Add selector API to Study permissions

The selector for a permission record will enable the signal for
study_permission_change to indicate the target for a permission, without
referencing database objects directly.

wmorrell committed:
Add API to fetch current user

EDD stores the request object in a threadlocal variable while processing
requests, so it can be available as a "global" variable in contexts where the
request is not explicitly passed as a parameter. This API fetches the user on
the global request, for times when code needs the user, but not necessarily
the request.

wmorrell committed:
Refactor variable naming in exports

Replacing nested attribute calls for finding the first study listed in an
export selection, with an explicit variable named `first_study`. This makes the
code a bit clearer on what it's doing, and will assist with sending signals for
when an export runs.

wmorrell committed:
Remove unused attribute definition

wmorrell committed:
Fix Update object(s) tied to wizard import

Every single record touched by the wizard importer would end up getting its own
unique Update record, rather than all items associated with a single import
sharing a single Update. Fixing it so that before records get created, a fake
"request" object with an Update for the entire import session is created. With
the fake request in place, all records resulting from the import session will
share one Update record.

wmorrell committed:
Add send calls for metrics signals

Inserting calls to the metrics signals in appropriate places in the EDD code.
EDD will now generate signals for:

-   study_created
-   study_described
-   study_exported
-   study_imported
-   study_permission_change
-   study_viewed
-   study_worklist

wmorrell commented:
Please update this by rebasing on latest `trunk`, and consolidating all the "wip" commits to what you believe is the final code to merge. It's hard to tell if you've actually resolved many of the comments I have made so far.

wmorrell committed:
Add database logging for study cycle metrics

Hooking up the signals for various study actions, to a database table to record
those actions. Includes an admin view that allows read-only listing of events.

wmorrell committed:
Update pytest warnings configs

Always set pytest to emit warnings with "default" strategy, of once per call
site. Ignore warnings from OpenPyXL re: ignoring unsupported extensions used
within an XLSX file.

Note: this has nothing to do with the filename extension. It's more-or-less
plugins used by Excel, that are not supported in the library. Since the library
does not understand the data provided by this "extension", it's warning that
the data for that extension is ignored.

wmorrell committed:
Replace uses of ICE Registry APIs

Deleting the old ICE APIs, and instead using the simplified API from the prior
commit. This is a big reduction in complexity, around a thousand lines of code
removed, and hundreds of statements and branches eliminated. It's also removing
the last of the warnings about hanging sockets after running full tests.

wmorrell committed:
Refactor registration of Celery task serializer

Moving the registration of our custom JSON serialization classes to just prior
to the initialization of the Celery app. Delaying the registration until the
`edd.utilities` module is imported would sometimes result in the serializers
not getting registered in time for use by Celery. The problem goes away if it
becomes impossible to initialize Celery before the registration.

wmorrell committed:
Add monkey patch for default requests timeout

The default timeout was previously applied to a custom subclass of the requests
Session class, which was only used by the code interfacing with the ICE REST
API. The updated monkey patch involves no subclassing, and merely wraps the
function that ultimately sends requests to include a default timeout
if-and-only-if a timeout was not already specified.

wmorrell committed:
Remove unused test code re: attachments

The test setup used a superfluous setUpTestData method, which set a confusing
URL for the test. The URL would later get overridden with the correct URL in
the test setUp method.

wmorrell committed:
Fix user detection outside of request

Previous implementation would raise an Exception outside the context of a
request, as "current" would not be defined. The function will now correctly
return the system user if no current request is in progress.

wmorrell committed:
Add simple export/worklist task tests

Basic tests for exports and worklists Celery tasks. Ensures coverage for the
new StudyLog records on all signals, prior to adding additional tests to check
that the records are added to the database.

wmorrell committed:
Add tests of appending StudyLog entries

The added code already had close to full coverage, because tests in other
modules were already causing the code to get executed. Adding some explicit
tests for showing that specific actions will result in specific StudyLog
entries added to the database.

In the process of adding tests, also filled in missing functionality from the
first pass at the feature. In particular, the signals for describing a Study
were not getting triggered via experiment description file uploads, nor for
combinatorial line creation UI.

All the other StudyLog entries record a count of the number of lines involved
in the entry (unit operations in the milestone metric terminology). However,
the exports do not operate directly on Study or Line objects, and only directly
reference the MeasurementValue records that get filtered through the export
request filters. To get this information involves two more queries to the
database as part of the exports.

Some small refactorings done with this work:

The REST views module was beginning to get fairly long, with both View/ViewSets
and FilterSets in the same file. Spinning out the FilterSets into their own
filters module, instead.

Also consolidating the GraphQL modules, which made use of the REST filters. It
made no sense to keep one in the "main" module, only to have it subclassed by
the real one in "edd" module.

wmorrell committed:
Rebasing master onto import-ambr-data

Remove defaults for LBL LDAP logins

The LDAP configuration should be specific to each deployment. Removing the
defaults that will always connect to LBL's LDAP servers unless explicitly
overridden. We should not be "hard-coding" this.

somtirtharoy committed:
Add mapper for measurement types to units

- Add a class DefaultUnit that maps measurement tyeps to corresponding units
- Added admin class to configure the model from the admin panel
- optimized code to use the class instead of hardcoded map
- optmized code to create instance of Ambr data only once
- enable test to query database

somtirtharoy committed:
Populate test databse with dump

- added dump from databse for the DefaultUnit table
- use the json file to load the data to populate the test database before running test for Ambr

wmorrell committed:
Change test deploy target to swarm-stage

somtirtharoy committed:
Update Pipfile.lock

wmorrell commented on server/edd/load/parsers/core.py:
```diff
@@ -641,6 +645,127 @@ def _raw_cell_value(self, cell):
+                two_cols = two_cols.dropna()
+                if not two_cols.dropna().empty:
+                    # decimate the data
+                    two_cols = two_cols.iloc[::10, :]
```
This does not belong in the parser stage. The parser should read *all* of the data, then pass it along. The code that converts parsed and resolved data into database records should make the determination of which items to drop, and whether to do anything more complicated than a blind, one-in-ten sampling.

wmorrell commented on server/edd/load/parsers/core.py:
```diff
@@ -641,6 +645,127 @@ def _raw_cell_value(self, cell):
+        self.df = df
+        # maps naming format of Ambr measurement type to
+        # naming convention in EDD
+        self.mtypes = {
```
Don't hard-code these. If the input file uses different names than what is currently in EDD, that translation needs to happen at the data level, and not as hard-coded entries.

wmorrell commented on server/main/models/core.py:
```diff
@@ -1375,3 +1375,27 @@ def active_in(cls, *, study_id, protocol_id, assay_id=None, id_range=None):
         return active
+
+
+class DefaultUnit(models.Model):
```
This model belongs in the `edd.load` module, not in `main`. It's specific to loading/importing data, not the core data models.

wmorrell commented on server/edd/load/tests/files/defaultunit_data.json:
```diff
@@ -0,0 +1 @@
+[{"model": "main.defaultunit", "pk": 5, "fields": {"measurement_type": 1, "unit": 1, "protocol": null, "parser": ""}}, {"model": "main.defaultunit", "pk": 6, "fields": {"measurement_type": 2, "unit": 17, "protocol": 8, "parser": "Ambr"}}, {"model": "main.defaultunit", "pk": 7, "fields": {"measurement_type": 4, "unit": 1, "protocol": null, "parser": ""}}, {"model": "main.defaultunit", "pk": 8, "fi ...
```
Should use a Factory for generating these in the tests that need them. Using a fixture file will hard-code ID values that may not be correct as the code changes.

somtirtharoy committed:
Make changes according to feedback to pull request

- remove instance of df_parsed form code
- move DefaultUnit model from main module to edd.load module
- restructured code to move data manipultion code out of the parse function

somtirtharoy committed:
Configure test database using factories

- Add factories for measurement types and units
- Add test setup to create the default units in the test database
- make changes in parser to persist changes in the parsed result

somtirtharoy committed:
Add data configuration to map input measurement names to EDD names

- Add model to map any input measuremetn names to corresponding EDD names
- Add admin class to configure the model through the admin panel
- Add factory class to populate it in test databse before testing code
- Modify parser code to use the model rther than use hardcoded data
- Create migrations
- Delete defaultunit_data.json file from tests

wmorrell commented:
Som,

You've added 117 commits to the pull request. You're re-writing a ton of stuff from EDD's history, and reverting quite a few things. Please fix your branch, and force push again.

somtirtharoy commented:
Hi William, I rebased master on my feature branch and pushed it to remote.
There were conflicts but I picked the changes from the trunk so that I do
not revert anything. I want to confirm that's what you want me to do?

Thanks,

On Tue, Oct 5, 2021 at 1:38 PM wmorrell ***@***.***> wrote:

> Som,
>
> You've added 117 commits to the pull request. You're re-writing a ton of
> stuff from EDD's history, and reverting quite a few things. Please fix your
> branch, and force push again.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JBEI/edd/pull/21#issuecomment-934795404>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKDGPCAGAVCQWD3LG7ZWGYTUFNO35ANCNFSM43PCKN7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

wmorrell commented:
No, that's precisely the *opposite* of what you should have done. You rebase *your* branch onto trunk, not trunk onto your code.

somtirtharoy commented:
Oops, okay I will work on it. I didn't want the trunk to be affected.
Thanks. I will have to undo the rebase and then do that again.

Thanks

On Tue, Oct 5, 2021 at 1:51 PM wmorrell ***@***.***> wrote:

> No, that's precisely the *opposite* of what you should have done. You
> rebase *your* branch onto trunk, not trunk onto your code.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JBEI/edd/pull/21#issuecomment-934814486>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKDGPCGOS25U3INVYB5SYXLUFNQNZANCNFSM43PCKN7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

somtirtharoy commented:
Hi William, I had to cherry pick my commits and then I rebased it onto the
trunk on my repo and created a new pull request. Please let me know if
everything looks alright and it can be pushed to production.

Thanks,

On Tue, Oct 5, 2021 at 1:53 PM Somtirtha Roy ***@***.***> wrote:

> Oops, okay I will work on it. I didn't want the trunk to be affected.
> Thanks. I will have to undo the rebase and then do that again.
>
> Thanks
>
> On Tue, Oct 5, 2021 at 1:51 PM wmorrell ***@***.***> wrote:
>
>> No, that's precisely the *opposite* of what you should have done. You
>> rebase *your* branch onto trunk, not trunk onto your code.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/JBEI/edd/pull/21#issuecomment-934814486>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AKDGPCGOS25U3INVYB5SYXLUFNQNZANCNFSM43PCKN7Q>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>>
>

wmorrell commented:
Closing this one in favor of the non-draft pull request. Almost there …
